src/Microsoft.Health.Fhir.CosmosDb/Features/Search/FhirCosmosSearchService.cs

```diff
@@ -429,13 +429,34 @@ private async Task<Expression> RecurseChainedExpression(ChainedExpression expres
 
             try
             {
-                (IReadOnlyList<T> results, string nextContinuationToken) = await _fhirDataStore.ExecuteDocumentQueryAsync<T>(sqlQuerySpec, feedOptions, continuationToken, searchOptions.MaxItemCountSpecifiedByClient, feedOptions.MaxConcurrency == null ? searchEnumerationTimeoutOverrideIfSequential : null, cancellationToken);
+                (IReadOnlyList<T> results, string nextContinuationToken) = (null, null);
+                var desiredItemCount = feedOptions.MaxItemCount * CosmosFhirDataStore.ExecuteDocumentQueryAsyncMinimumFillFactor;
 
-                if (queryPartitionStatistics != null && messagesList != null)
+                // Set timeout for sequential query execution
+                if (feedOptions.MaxConcurrency == null &&
+                    _cosmosConfig.ParallelQueryOptions.EnableConcurrencyIfQueryExceedsTimeLimit == true &&
+                    string.IsNullOrEmpty(searchOptions.ContinuationToken))
                 {
-                    var desiredItemCount = feedOptions.MaxItemCount * CosmosFhirDataStore.ExecuteDocumentQueryAsyncMinimumFillFactor;
+                    searchEnumerationTimeoutOverrideIfSequential = TimeSpan.FromSeconds(5);
+
+                    // Executing query sequentially until the timeout
+                    (results, nextContinuationToken) = await _fhirDataStore.ExecuteDocumentQueryAsync<T>(sqlQuerySpec, feedOptions, continuationToken, searchOptions.MaxItemCountSpecifiedByClient, searchEnumerationTimeoutOverrideIfSequential, cancellationToken);
 
+                    // check if we need to restart the query in parallel
                     if (results.Count < desiredItemCount && !string.IsNullOrEmpty(nextContinuationToken))
+                    {
+                        feedOptions.MaxConcurrency = _cosmosConfig.ParallelQueryOptions.MaxQueryConcurrency;
+                        (results, nextContinuationToken) = await _fhirDataStore.ExecuteDocumentQueryAsync<T>(sqlQuerySpec, feedOptions, continuationToken, searchOptions.MaxItemCountSpecifiedByClient, null, cancellationToken);
```

Comment:
Observations:

1. We try sequentially for 5 seconds and give up
2. We reissue the query with MaxParallelism, resetting the initial resultset

In the case that (2) doesn't return the desired number of results, what happens in the cases of:

- The execution with MaxParallelism returns less than the desired number of results 
- If the first statement had `searchOptions.MaxItemCountSpecifiedByClient = true` and used the previous page's continuation token, would we execute in parallel unnecessary?

Reply:
I think in this case searchOptions.ContinuationToken would not be null, so it won't be re-executed.

Reply:
```suggestion
                        (results, nextContinuationToken) = await _fhirDataStore.ExecuteDocumentQueryAsync<T>(sqlQuerySpec, feedOptions, null, searchOptions.MaxItemCountSpecifiedByClient, null, cancellationToken);
```

Reply:
Thanks! Done